File: Algorithms/1190_Reverse_Substrings_Between_Each_Pair_of_Parentheses/Solution.kt
import java.util.*

class Solution {
    fun reverseParentheses(s: String): String = Stack<String>().apply {
        s.forEach {
            this.push(
                if (it == ')') this.pops()
                else it.toString()
            )
        }
    }.pops()

    private fun Stack<String>.pops(): String = StringBuilder().also {
        while (this.isNotEmpty()) {
            val top = this.pop()
            if (top != "(") it.append(top) else break
        }
    }.reverse().toString()
}

File: Algorithms/1190_Reverse_Substrings_Between_Each_Pair_of_Parentheses/solution.py
from typing import *


class Solution:
    def reverseParentheses(self, s: str) -> str:
        stk: List[str] = []

        for c in s:
            stk.append(self.pops(stk) if c == ')' else c)

        return self.pops(stk)

    @staticmethod
    def pops(stk: List[str]) -> str:
        s: List[str] = []
        while stk:
            top = stk.pop()
            if top != '(':
                s.append(top)
            else:
                break
        return ''.join(s)[::-1]
